test/mpi/comm/idup_with_info.c

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpi.h"
+#include <stdio.h>
+#include "mpitest.h"
+
+int run_tests(MPI_Comm comm)
+{
+    int rank, size, wrank, wsize, dest, a, b, errs = 0;
+    MPI_Status status;
+
+    /* Check basic properties */
+    MPI_Comm_size(MPI_COMM_WORLD, &wsize);
+    MPI_Comm_rank(MPI_COMM_WORLD, &wrank);
+    MPI_Comm_size(comm, &size);
+    MPI_Comm_rank(comm, &rank);
+
+    if (size != wsize || rank != wrank) {
+        errs++;
+        fprintf(stderr, "Size (%d) or rank (%d) wrong\n", size, rank);
+        fflush(stderr);
+    }
+
+    MPI_Barrier(comm);
+
+    /* Can we communicate with this new communicator? */
+    dest = MPI_PROC_NULL;
+    if (rank == 0) {
+        dest = size - 1;
+        a = rank;
+        b = -1;
+        MPI_Sendrecv(&a, 1, MPI_INT, dest, 0, &b, 1, MPI_INT, dest, 0, comm, &status);
+        if (b != dest) {
+            errs++;
+            fprintf(stderr, "Received %d expected %d on %d\n", b, dest, rank);
+            fflush(stderr);
+        }
+        if (status.MPI_SOURCE != dest) {
+            errs++;
+            fprintf(stderr, "Source not set correctly in status on %d\n", rank);
+            fflush(stderr);
+        }
+    } else if (rank == size - 1) {
+        dest = 0;
+        a = rank;
+        b = -1;
+        MPI_Sendrecv(&a, 1, MPI_INT, dest, 0, &b, 1, MPI_INT, dest, 0, comm, &status);
+        if (b != dest) {
+            errs++;
+            fprintf(stderr, "Received %d expected %d on %d\n", b, dest, rank);
+            fflush(stderr);
+        }
+        if (status.MPI_SOURCE != dest) {
+            errs++;
+            fprintf(stderr, "Source not set correctly in status on %d\n", rank);
+            fflush(stderr);
+        }
+    }
+
+    MPI_Barrier(comm);
+
+    return errs;
+}
+
+int main(int argc, char **argv)
+{
+    int errs = 0;
+    MPI_Comm newcomm1, newcomm2, newcomm3, newcomm4;
+    MPI_Request reqs[2];
+    MPI_Info info;
+    int rank;
+
+    MTest_Init(&argc, &argv);
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+    /* Dup with no info */
+    MPIX_Comm_idup_with_info(MPI_COMM_WORLD, MPI_INFO_NULL, &newcomm1, &reqs[0]);
+    MPI_Wait(&reqs[0], MPI_STATUS_IGNORE);
+    errs += run_tests(newcomm1);
+
+    MPIX_Comm_idup_with_info(MPI_COMM_WORLD, MPI_INFO_NULL, &newcomm2, &reqs[0]);
+    MPI_Wait(&reqs[0], MPI_STATUS_IGNORE);
+    errs += run_tests(newcomm2);
+
+    /* Dup with info keys in mixed order */
+    MPI_Info_create(&info);
+    MPI_Info_set(info, (char *) "host", (char *) "myhost.myorg.org");
```

Comment:
should we check to see if these info are set on the new communicators?

Reply:
It's hard to check unless we make the test mpich specific.  Other MPI implementations are not required to set them.  But I can add another test in the impls directory to check the info keys.

Reply:
Ah, right. This is OK then.